src/overlap_pmmg.c

```diff
@@ -0,0 +1,736 @@
+     *     (3) k: position of node ip in external communicator Pcolor_out-Pcolor_ter
+     *     (4) ip_in: local index of node on Pcolor_in
+     *     (5) ip_out: local index of node on Pcolor_out **/
+    mesh->xpmax = MG_MAX( (long long)(1.5*mesh->xp),mesh->npmax);
+
+    /* Loop over the points having MG_PARBDY tag */
+    for (i=0; i < npPBDY_out2in; i++) {
+      ip_out   = pointIdxPBDY_ToRecv[i];
+      coord[0] = pointCoordPBDY_ToRecv[3*i];
+      coord[1] = pointCoordPBDY_ToRecv[3*i+1];
+      coord[2] = pointCoordPBDY_ToRecv[3*i+2];
+      tag      = pointTagPBDY_ToRecv[i]|MG_OVERLAP; // Add the tag MG_OVERLAP to this point
+      ref      = pointRefPBDY_ToRecv[i];
+      ls_val   = lsPBDY_ToRecv[i];
+
+      /* Loop over the data allowing to identify these points */
+      for (j=0; j < ndataPBDY_out2in; j++) {
+        duplicated_point = 0;
+
+        color_ter = dataPBDY_ToRecv[3*j];
+        k         = dataPBDY_ToRecv[3*j+1]; // Position in external communicator Pcolor_out-Pcolor_ter
+        ip        = dataPBDY_ToRecv[3*j+2]; // Index of point on Pcolor_out
+
+        if ( !(ip == ip_out) ) continue;
+
+        /* Search into dataPBDY_AlreadyAdded to seen if it exists already */
+        for (r=0; r < ndataPBDY_added+1; r++) {
+          if ( (color_out == dataPBDY_AlreadyAdded[5*r+1]) && (color_ter == dataPBDY_AlreadyAdded[5*r])) {
+            if ( k == dataPBDY_AlreadyAdded[5*r+2] ) {
+              ip_in = dataPBDY_AlreadyAdded[5*r+3];
+              duplicated_point =1;
+              break;
+            }
+          }
+          if ( (color_out == dataPBDY_AlreadyAdded[5*r]) && (ip_out == dataPBDY_AlreadyAdded[5*r+4])) {
+              ip_in = dataPBDY_AlreadyAdded[5*r+3];
+              duplicated_point =1;
+              break;
+          }
+        }
+
+        /* If this point has not been added yet, then add it to the mesh */
+        if (duplicated_point==0) {
+          if ( color_ter != color_in ) {
+            ip_in = MMG3D_newPt(mesh,coord,tag,1);
+            mesh->point[ip_in].ref = ref; // Add ref
+            mesh->point[ip_in].xp  = 0;   // Assign 0 to xp
+            ls->m[ip_in]           = ls_val;
+          }
+        }
+
+        /* Update the hash tables to know correspondance of local index on color_in and color_out */
+        hash_out2in[ip_out] = ip_in;  // From index on color_out, I found index on color_in
+        hash_in2out[ip_in]  = ip_out; // From index on color_in,  I found index on color_out
+
+        /* Update dataPBDY_AlreadyAdded data necessary to identify which MG_PARBDY points have already been treated */
+        dataPBDY_AlreadyAdded[5*ndataPBDY_added]  = color_out;
+        dataPBDY_AlreadyAdded[5*ndataPBDY_added+1]= color_ter;
+        dataPBDY_AlreadyAdded[5*ndataPBDY_added+2]= k;
+        dataPBDY_AlreadyAdded[5*ndataPBDY_added+3]= ip_in;
+        dataPBDY_AlreadyAdded[5*ndataPBDY_added+4]= ip_out;
+        ndataPBDY_added++;
+      }
+    }
```

Comment:
All the previous loop seems pretty complex and hard to understand : I wonder if it can be simplified or how to improve the comments ?

Reply:
I have tried to improve the comments a lot - hopefully this will help clarifying things, otherwise let me know.